server-spi/src/main/java/org/keycloak/models/credential/RecoveryAuthnCodesCredentialModel.java

```diff
@@ -0,0 +1,106 @@
+            RecoveryAuthnCodesSecretData secretData) {
+        this.credentialData = credentialData;
+        this.secretData = secretData;
+    }
+
+    public RecoveryAuthnCodeRepresentation getNextRecoveryAuthnCode() {
+        return this.secretData.getCodes().get(0);
+    }
+
+    public boolean allCodesUsed() {
+        return this.secretData.getCodes().isEmpty();
+    }
+
+    public void removeRecoveryAuthnCode() {
+        try {
+            this.secretData.removeNextBackupCode();
+            this.credentialData.setRemainingCodes(this.secretData.getCodes().size());
+            this.setSecretData(JsonSerialization.writeValueAsString(this.secretData));
+            this.setCredentialData(JsonSerialization.writeValueAsString(this.credentialData));
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public static RecoveryAuthnCodesCredentialModel createFromValues(List<String> originalGeneratedCodes, long generatedAt,
+            String userLabel) {
+        RecoveryAuthnCodesSecretData secretData;
+        RecoveryAuthnCodesCredentialData credentialData;
+        RecoveryAuthnCodesCredentialModel model;
+
+        try {
+            List<RecoveryAuthnCodeRepresentation> recoveryCodes = IntStream.range(0, originalGeneratedCodes.size())
+                    .mapToObj(i -> new RecoveryAuthnCodeRepresentation(i + 1,
+                            RecoveryAuthnCodesUtils.hashRawCode(originalGeneratedCodes.get(i))))
+                    .collect(Collectors.toList());
+            secretData = new RecoveryAuthnCodesSecretData(recoveryCodes);
+            credentialData = new RecoveryAuthnCodesCredentialData(RecoveryAuthnCodesUtils.NUM_HASH_ITERATIONS,
+                    RecoveryAuthnCodesUtils.NOM_ALGORITHM_TO_HASH, recoveryCodes.size(), recoveryCodes.size());
+            model = new RecoveryAuthnCodesCredentialModel(credentialData, secretData);
+            model.setCredentialData(JsonSerialization.writeValueAsString(credentialData));
+            model.setSecretData(JsonSerialization.writeValueAsString(secretData));
+            model.setCreatedDate(generatedAt);
+            model.setType(TYPE);
+
+            if (userLabel != null) {
+                model.setUserLabel(userLabel);
+            }
+            return model;
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public static RecoveryAuthnCodesCredentialModel createFromCredentialModel(CredentialModel credentialModel) {
+        RecoveryAuthnCodesCredentialData credentialData;
+        RecoveryAuthnCodesSecretData secretData = null;
+        RecoveryAuthnCodesCredentialModel newModel;
+        try {
+            credentialData = JsonSerialization.readValue(credentialModel.getCredentialData(),
+                    RecoveryAuthnCodesCredentialData.class);
+            if (credentialModel.getSecretData() != null) {
```

Comment:
Since the credentialModel may have the secret removed for security reasons, I'm having to add this guard clause.

Reply:
This pattern seems to be used multiple times - perhaps a common helper method could help to remove some code duplication.

Reply:
I believe I can remove this if-statement completely.